app/models/assignment.rb

```diff
@@ -605,6 +605,65 @@ def summary_json(user)
       numMarked: self.get_num_marked(user.admin? ? nil : user.id) }
   end
 
+  # Generates the summary of the most test results associated with an assignment.
+  def summary_test_results
+    test_groups_query = self.test_groups
+                            .joins(test_group_results: { test_run: :grouping })
+                            .group('test_groups.id', 'groupings.id')
+                            .select('test_groups.id AS test_groups_id',
+                                    'MAX(test_group_results.created_at) AS test_group_results_created_at')
+                            .where.not('test_runs.submission_id': nil)
+                            .to_sql
+
+    TestGroup.joins(test_group_results: [:test_results, { test_run: { grouping: :group } }])
+             .joins("INNER JOIN (#{test_groups_query}) sub \
+                     ON test_groups.id = sub.test_groups_id \
+                     AND test_group_results_test_groups.created_at = sub.test_group_results_created_at")
+             .select('test_groups.name',
+                     'test_groups_id',
+                     'groups.group_name',
+                     'test_results.name as test_result_name',
+                     'test_results.status',
+                     'test_results.marks_earned',
+                     'test_results.marks_total',
+                     :output, :extra_info, :error_type)
+  end
+
+  # Generate a JSON summary of the most recent test results associated with an assignment.
+  def summary_test_result_json
+    self.summary_test_results.group_by(&:group_name).transform_values do |grouping|
+      grouping.group_by(&:name)
+    end.to_json
+  end
+
+  # Generate a CSV summary of the most recent test results associated with an assignment.
+  def summary_test_result_csv
+    summary_test_results = self.summary_test_results
+    results = summary_test_results.map do |result|
+      ["#{result['group_name']}:#{result['test_result_name']}", result['status']]
```

Comment:
You're still confusing `group_name` with `test_groups.name`.

You need `test_groups.name` here, since that's the actual name of the test group. 

`group_name` is different, it identifies the student group. Overall you need a *nested hash* so that the outer hash keys are the group names, and the inner hash keys are what you had before (the test group name and then the test name).

This way, you can take then do something like `results[some_group_name][header]` to actually get the status.